src/main/java/io/jenkins/plugins/checks/status/BuildStatusChecksPublisher.java

```diff
@@ -58,6 +77,207 @@ private static AbstractStatusChecksProperties findProperties(final Job<?, ?> job
+            }
+        }
+        return null;
+    }
+
+    private static String extractOutputTitle(final Run run) {
+        Result result = run.getResult();
+        if (result == null) {
+            return "In progress";
+        }
+        if (result.isBetterOrEqualTo(Result.SUCCESS)) {
+            return "Success";
+        }
+        if (result.isBetterOrEqualTo(Result.UNSTABLE)) {
+            return "Unstable";
+        }
+        if (result.isBetterOrEqualTo(Result.FAILURE)) {
+            return "Failure";
+        }
+        if (result.isBetterOrEqualTo(Result.NOT_BUILT)) {
+            return "Skipped";
+        }
+        if (result.isBetterOrEqualTo(Result.ABORTED)) {
+            return "Aborted";
+        }
+        throw new IllegalStateException("Unsupported run result: " + result);
+    }
+
+    @SuppressWarnings({"PMD.ConfusingTernary", "PMD.NPathComplexity", "JavaNCSS"})
+    static ChecksOutput getOutput(final Run run, final FlowExecution execution) {
+
+        FlowGraphTable table = new FlowGraphTable(execution);
+        table.build();
+
+        Stack<Integer> indentationStack = new Stack<>();
+
+        TruncatedStringBuilder summaryBuilder = new TruncatedStringBuilder(MAX_MSG_SIZE_TO_CHECKS_API, TRUNCATED_MESSAGE);
+        TruncatedStringBuilder textBuilder = new TruncatedStringBuilder(MAX_MSG_SIZE_TO_CHECKS_API, TRUNCATED_MESSAGE);
+
+        table.getRows().forEach(row -> {
+            final FlowNode flowNode = row.getNode();
+
+            boolean isStage = isStage(flowNode);
+            boolean isParallel = isParallelBranch(flowNode);
+            ErrorAction errorAction = flowNode.getError();
+            WarningAction warningAction = flowNode.getPersistentAction(WarningAction.class);
+
+            if (!isStage
+                    && !isParallel
+                    && errorAction == null
+                    && warningAction == null) {
+                return;
+            }
+
+            StringBuilder nodeSummaryBuilder = new StringBuilder();
+            StringBuilder nodeTextBuilder = new StringBuilder();
+
+            if (isStage || isParallel) {
+                while (!indentationStack.isEmpty() && row.getTreeDepth() < indentationStack.peek()) {
+                    indentationStack.pop();
+                }
+                if (indentationStack.isEmpty() || row.getTreeDepth() > indentationStack.peek()) {
+                    indentationStack.push(row.getTreeDepth());
+                }
+                nodeTextBuilder.append(String.join("", Collections.nCopies(indentationStack.size(), "  ")));
+                nodeTextBuilder.append("* ");
+
+                final String displayName;
+                // Duplicate parallel detection logic to keep spotbugs happy :(
+                ThreadNameAction threadNameAction = flowNode.getAction(ThreadNameAction.class);
+                if (threadNameAction != null) {
+                    displayName = threadNameAction.getThreadName();
```

Comment:
you could also do:
```suggestion   
                if (isParallel) {
                    ThreadNameAction threadNameAction = flowNode.getAction(ThreadNameAction.class);
                    // not possible to be null as checked earlier but spotbugs doesn't understand it
                    assert threadNameAction != null;
                    displayName = threadNameAction.getThreadName();
```

or in the isParallel check you could return the FlowNode, either by replacing the boolean or by using something like a `Pair`

Reply:
I've realised I don't care whether it's stage or branch any more, so I can simplify all the logic around this...